[RFC] DX11 several optimizations of copying frames to/from gpu.

afedchin opened the pull request:
Set of improvements. For more info please see commit messages.
1. CopyDXVA2Picture - here is a comparsion between different methods of copying dxva FHD frame into sysmem.
   ![gpucopy](https://cloud.githubusercontent.com/assets/2063885/8908853/0bfaf4bc-3487-11e5-8023-97995d6a5a6f.jpg)
2. CProcessorHD::Convert  - here is a comparsion between old memcpy and new sse2 method (used FHD frame).
   ![dxvacopy](https://cloud.githubusercontent.com/assets/2063885/8908884/28334af8-3487-11e5-8601-2d9ad16873e1.jpg)

@fritsch @FernetMenta please take a look.

Commit:
[dx11] Moved CDVDCodecUtils::CopyDXVA2Picture to WinRenderer and optimize copying a DXVA texture to YUV buffer with sse4.

Commit:
[dxva] CProcessorHD::Convert - Optimize method with sse2 instructions.

Commit:
[dx11] CProcessorHD: Get rig of unneeded std::map.

FernetMenta commented on xbmc/cores/VideoRenderers/WinRenderer.cpp:
```diff
@@ -1354,9 +1380,120 @@ void YUVBuffer::Clear()
+  return SUCCEEDED(hr);
+}
+
+void YUVBuffer::PerformCopy()
```
what is the benefit in moving this from decoder to render thread? I would consider all load taken off from the main thread a plus.

afedchin replied:
Agree with you. It would be excellent if you point me where I can do this into decoder thread. The main issue is we need to do actual copying as late as possible to give DX runtime make copy of frame in background.

FernetMenta replied:
I would have expected it in CopyFromDXVA. That is last task of decoder thread before hand-off.

afedchin replied:
I understand. Currently the CopySubresourceRegion is performed only on decoder thread. if you object to this change then I'll revert it and keep actual copying into decoder thread. so we have to accept the performance penalty of calling Map right after CopySubresourceRegion. Using sse4 give us a good benefit which covers this penalty.

FernetMenta replied:
No, I don't object. I was just asking. If you have good reason, go ahead.
fyi: in order to cope with that kind of problem vaapi has an embedded thread which does copy and/or texture work. maybe you want to do something similar in a future iteration.

afedchin replied:
I have no good reason to move this out from decoder thread also because it's used only if user defines <forcedxvarender>false</forcedxvarender> in as.xml. So I think it will be a good compromise to left it as is is and just update to using sse4.

Paxxi commented on xbmc/utils/win32/gpu_memcpy_sse4.h:
```diff
@@ -0,0 +1,128 @@
+    __m128i xmm8, xmm9, xmm10, xmm11, xmm12, xmm13, xmm14, xmm15;
+#endif
+
+    size_t reminder = size & (regsInLoop * sizeof(xmm0) - 1); // Copy 128 or 256 bytes every loop
```
Just a notice, Should be remainder

afedchin replied:
It's a original design. I didn't change it.

FernetMenta commented on xbmc/utils/win32/gpu_memcpy_sse4.h:
```diff
@@ -0,0 +1,128 @@
+// GPU tiled memory (write back)
+// Performance tip: page offset (12 lsb) of both addresses should be different
+//  optimally use a 2K offset between them.
+inline void* gpu_memcpy(void* d, const void* s, size_t size)
```
On Linux this would prevent the application from loading. For this reason I put the sse4 code into a dll and only load this when sse4 is detected to be available.
https://github.com/xbmc/xbmc/tree/master/xbmc/linux/sse4
Are you sure Windows behaves differently? Did you test on systems without sse4?

afedchin replied:
Unfortunately I have no system without sse4. I'm not be able to test this. But this was taken from LAV filters project which is compiled with this and works fine. Also msvc has no flag to enable sse4 intrinsics like gcc has.

fritsch commented on xbmc/utils/win32/gpu_memcpy_sse4.h:
```diff
@@ -0,0 +1,128 @@
+        xmm13 = _mm_stream_load_si128(pSrc + 13);
+        xmm14 = _mm_stream_load_si128(pSrc + 14);
+        xmm15 = _mm_stream_load_si128(pSrc + 15);
+#endif
```
I fully miss the fence here ... how should the store guarantee that all of those above are loaded and synced?

afedchin replied:
here you go https://github.com/xbmc/xbmc/pull/7637/files#diff-6584b1c7a5567de16d7e4e1ef0b302d9R54

fritsch replied:
from my understanding that needs to happen after the loads, so that the store reads the correct data. Am I wrong here?

fritsch replied:
See the original: https://software.intel.com/sites/default/files/m/d/4/1/d/8/FastCopy_of_Decoded_Video_Frames.pdf

afedchin replied:
Intel's implementation uses cache block and two different loops to write into cache and then write back into memory. It requires to be sure what all loads and stores are completed between loops. This is why it requires fence between loops. This implementation uses only one loop and it no needs to fence in loop. Correct me if I'm wrong here.

fritsch replied:
That's exactly the question :-)

fritsch replied:
@Voyager1 we need your help here!

Voyager1 replied:
I'd say we only need to fence if we write to the same memory block several times potentially in out of place order (like in Intel's example, the cache block). Having one fence before this loop seems to be enough. 
Take a look at the Intel document though, Section 3.1.2 and Table 3-1. It's for a reason that they use the cache block. 
That said, this one seems to be written as a replacement for memcpy and is not necessarily used to copy an entire frame. Not everytime a cache block may be appropriate.
Besides a suboptimal performance (if you read the theory on streaming fill buffers in the CPU, section 3.1.1) I don't see anything wrong and the performance, although suboptimal, seems to be way better than memcpy...
Sorry for the long winded answer, but I'd say it's good as it is.

fritsch replied:
Thanks much. Then it can go in as is!

fritsch replied:
Yeah on linux we use that 4k block out of performance reason - I asked @afedchin before why he did not use it - but forgot the answer to be honest :-)

afedchin replied:
thanks for detailed explanation.

afedchin replied:
@fritsch I've tried it. But seems it's a little slower than Hendrix's version. (look the comparison at the opening this PR)

fritsch replied:
I see, yes. Perhaps the figures changes under more load - but it's fine as is.

a11599 commented:
Does it make sense to keep the AS setting with this speed improvement? We could maybe honor the GUI setting for PS render method now. Those who want DXVA rendering for HW decoded video can still use Auto render method.

afedchin commented:
@FernetMenta I've updated this. moved the copying frame back to a decoder thread. please take a look

@a11599 I'm not sure. We get a real benefit on sse4 systems only. But at the other side, porting to dx11 already drops supporting some very old and legacy GPUs. Those who want PS rendering with hw decoder already knows about AS setting. Also take in account what we enabled HQ scalers with dxva rendering already by-default. Which benefit of PS rendering? Need to discuss this.

a11599 commented:
Actually the benefit I see from PS rendering over DXVA is that the D3D11 video processor will not do something behind your back with the image. Good for purists. Also currently both PS and Auto render method do the same without the AS making the two options redundant. But I do not want to hijack this PR so sorry for the noise.

afedchin commented:
@a11599 I need your opinion also. Let's start a thread at the forum (perhaps public) to discuss that. What do you think?

a11599 commented:
For now I think this is good as-is. Maybe later if (ever) our PS renderer gets some quality improvements like dithering we migth rethink the AS. Would be nice to test on a non-sse4 cpu but I do not have any. Maybe we can provide some test build and ask for testers on the forum?

FernetMenta commented:
@afedchin sorry, I almost missed that you pinged me. You are the expert here. From what I can tell it looks ok.

afedchin commented:
updated and squashed. jenkins build this please

MartijnKaijser commented:
jenkins build this please

hudokkow commented:
@afedchin please rebase

afedchin commented:
jenkins build this please

afedchin commented:
jenkins build and merge
